### PR TITLE
Add UpdateInvocation RPC for editing invocation permissions

### PR DESCRIPTION
**Pull request by bduffany**

### CODE REVIEW

File: server/buildbuddy_server/buildbuddy_server.go [hunk 1]
```diff
@@ -87,6 +89,60 @@ func (s *BuildBuddyServer) SearchInvocation(ctx context.Context, req *inpb.Searc
...
+		return status.UnimplementedError("Not Implemented")
+	}
+	if _, err := auth.AuthenticatedUser(ctx); err != nil {
+		return err
+	}
+
+	// Invocations should probably never be writeable by OTHERS,
+	// but this check is included for completeness.
+	if in.Perms&perms.OTHERS_WRITE != 0 {
+		return nil
```
Comment (tylerwilliams) [inline server/buildbuddy_server/buildbuddy_server.go:104]: as written, this silently fails -- do you want to log something when this happens?

Maybe

log.Print("Ignoring request to allow OTHERS_WRITE on invocation. This should not happen!")

**Comment by bduffany** [comment]
Can probably hold off on review until https://github.com/buildbuddy-io/buildbuddy/pull/109 is submitted since we want to check whether sharing is enabled as part of this RPC impl. If sharing is disabled then permissions should not be editable.

File: server/buildbuddy_server/buildbuddy_server.go [hunk 2]
```diff
@@ -87,6 +89,60 @@ func (s *BuildBuddyServer) SearchInvocation(ctx context.Context, req *inpb.Searc
...
+	in, err := db.LookupInvocation(ctx, req.GetInvocationId())
+	if err != nil {
+		return nil, err
+	}
+	if err := s.authorizeInvocationUpdate(ctx, req.GetRequestContext(), in); err != nil {
+		return nil, err
+	}
+	// NOTE: We currently ignore the user_id and group_id in the updated ACL.
+	in.Perms = updatedPerms
+	if err := db.InsertOrUpdateInvocation(ctx, in); err != nil {
```
Comment (tylerwilliams) [inline server/buildbuddy_server/buildbuddy_server.go:140]: One slight problem with the way this is currently written is that if an invocation is in-flight (picture a multi-hour long test run), and a user updates permissions on it, it's possible you'll 1) read data at t=0, 2) a invocation update will happen at t=1, and 3) you'll write the old data from t=0 back to the DB with the new ACLs at t=2.

One simple fix would be to only modify the ACL column when you update ACLs? And even better is doing this in a transaction so you're sure the modification happens atomically. I know this is probably a pain -- lmk if you want to chat about the best way to fix it.

**Review by tylerwilliams — CHANGES_REQUESTED** [review]

**Review by bduffany — COMMENTED** [review]
Made a few changes:

- Added `UpdateInvocationPermissions` to invocationdb.go, so that the update is performed in a transaction
- Extracted ACL checking logic to perms.go since we want to be able to reuse the logic for other ACLs.

File: server/backends/invocationdb/invocationdb.go [hunk 3]
```diff
@@ -61,6 +63,28 @@ func (d *InvocationDB) InsertOrUpdateInvocation(ctx context.Context, ti *tables.
...
 }
 
+func (d *InvocationDB) UpdateInvocationACL(ctx context.Context, invocationID string, acl *aclpb.ACL) error {
+	p, err := perms.ToPerms(acl)
+	if err != nil {
+		return err
+	}
+	return d.h.Transaction(func(tx *gorm.DB) error {
+		var in tables.Invocation
+		if err := tx.Select("perms", "user_id", "group_id").Where("invocation_id = ?", invocationID).First(&in).Error; err != nil {
```
Comment (tylerwilliams) [inline server/backends/invocationdb/invocationdb.go:73]: would prefer to use raw queries rather than gormy things -- shouldn't be much more code

**Review by tylerwilliams — CHANGES_REQUESTED** [review]
Thanks, this is much better! Few small things and we're good to go

File: server/util/perms/perms.go [hunk 4]
```diff
@@ -97,6 +99,39 @@ func ToPerms(acl *aclpb.ACL) (int, error) {
...
+
+	perms, err := ToPerms(acl)
+	if err != nil {
+		return err
+	}
+
+	if perms&OTHERS_WRITE != 0 {
+		log.Print("Ignoring request to allow OTHERS_WRITE. This should not happen!")
+	}
+	isOwner := reqCtx.GetUserId().GetId() == acl.GetUserId().GetId()
```
[the remark this answers is not in the document]
Comment (bduffany) [reply server/util/perms/perms.go:124]: Mind if we keep this for now to unblock this PR, and add a TODO instead? (Per offline discussion, AuthenticatedUser doesn't yet have GroupID populated with the selected group ID, and I think that might be a bit of an involved PR)

File: server/backends/invocationdb/invocationdb.go [hunk 5]
```diff
@@ -61,6 +63,28 @@ func (d *InvocationDB) InsertOrUpdateInvocation(ctx context.Context, ti *tables.
...
+			return err
+		}
+		currentACL := perms.ToACLProto(&uidpb.UserId{Id: in.UserID}, in.GroupID, in.Perms)
+		if err := perms.AuthorizeWrite(ctx, currentACL); err != nil {
+			return err
+		}
+		if err := tx.Exec(`
```
[the remark this answers is not in the document]
Comment (bduffany) [reply server/backends/invocationdb/invocationdb.go]: Done

File: server/backends/invocationdb/invocationdb.go (hunk 3 above)
Comment (bduffany) [reply server/backends/invocationdb/invocationdb.go]: Done

File: server/backends/invocationdb/invocationdb.go [hunk 6]
```diff
@@ -61,6 +63,35 @@ func (d *InvocationDB) InsertOrUpdateInvocation(ctx context.Context, ti *tables.
 	})
 }
 
+func (d *InvocationDB) UpdateInvocationACL(ctx context.Context, invocationID string, acl *aclpb.ACL) error {
+	p, err := perms.ToPerms(acl)
```
Comment (tylerwilliams) [inline server/backends/invocationdb/invocationdb.go:67]: I think perms.FromACL(acl) would be a clearer name here. wdyt?

**Review by tylerwilliams — CHANGES_REQUESTED** [review]
minor nit -- otherwise LGTM!

File: server/backends/invocationdb/invocationdb.go (hunk 6 above)
Comment (bduffany) [reply server/backends/invocationdb/invocationdb.go:67]: SG, Done

**Review by tylerwilliams — APPROVED** [review]
LGTM

**Pull request merged**